<a href="https://colab.research.google.com/github/nurfnick/Data_Viz/blob/main/19_Missing_and_Incomplete.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Missing and Incomplete

Often datasets will be missing entries.  There are many approaches we can take to dealing with these errors and omissions.  I will examine a dataset on the characters from <ins> The Lord of The Rings</ins>

In [17]:
import pandas as pa

df = pa.read_csv('https://raw.githubusercontent.com/nurfnick/Data_Viz/main/lotr_characters.csv')

df.head()

birth              death  gender  ...    race  realm        spouse
0      NaN                NaN  Female  ...     Men    NaN       Belemir
1  TA 2978  February 26 ,3019    Male  ...     Men    NaN           NaN
2      NaN        March ,3019    Male  ...    Orcs    NaN           NaN
3   TA 280             TA 515    Male  ...     Men  Arnor  Unnamed wife
4      NaN                NaN    Male  ...  Dragon    NaN           NaN

[5 rows x 9 columns]

We see right away that there are lots of `NaN`'s.  This is an empty field in our dataset.  Some characters are mentioned but never given much more background than a name.

In [18]:
df.isnull().sum(axis = 0)

birth     207
death     315
gender    143
hair      734
height    813
name        0
race      140
realm     714
spouse    403
dtype: int64

There are null values in every column except name.

In [26]:
df.isnull().sum(axis = 1).value_counts().sort_index()

0     15
1     59
2    185
3    236
4    178
5     81
6     20
7      1
8    136
dtype: int64

Here we see that there are only 15 entries with all fields and 136 that are name only (since name was never blank!)  Let's look at just those characters.

In [32]:
df[~df.isnull().any(axis = 1)]

birth  ...                               spouse
125                    SA 3209  ...                         Unnamed wife
134  YT, and perhaps firstborn  ...                         Unnamed wife
166                         YT  ...                              Aredhel
186                    TA 2990  ...  Lothíriel after the War of the Ring
194                     FA 532  ...                            Celebrían
204                    SA 3119  ...                         Unnamed wife
530                         YT  ...                            Galadriel
551     Possibly pre First Age  ...                     Most likely none
579                       3019  ...                                 None
620                    TA 2925  ...                         Unnamed wife
686                    YT 1362  ...                             Celeborn
692                    YT 1169  ...                             Nerdanel
795                  First Age  ...                         Unnamed wife
802                    YT 1050  ...                               Melian
873              March 1 ,2931  ...                                Arwen

[15 rows x 9 columns]